src/tracing/trace.ts

```diff
@@ -0,0 +1,59 @@
+import * as Sentry from '@sentry/react'
+import { Span, SpanStatusType } from '@sentry/tracing'
+
+interface TraceMetadata {
+  /** Arbitrary data stored on a trace. */
+  data?: Record<string, unknown>
+  /** Indexed (ie searchable) tags associated with a trace. */
+  tags?: Record<string, string | number | boolean>
+}
+
+// These methods are provided as an abstraction so that users will not interact with Sentry directly.
+// This avoids tightly coupling Sentry to our instrumentation outside of this file, in case we swap services.
+interface TraceCallbackOptions {
+  traceChild<T>(name: string, callback: TraceCallback<T>, metadata?: TraceMetadata): Promise<T>
+  setTraceData(key: string, value: unknown): void
+  setTraceTag(key: string, value: string | number | boolean): void
+  setTraceStatus(status: number | SpanStatusType): void
+  setTraceError(error: unknown): void
+}
+type TraceCallback<T> = (options: TraceCallbackOptions) => Promise<T>
+
+function traceTransaction(transaction?: Span) {
+  const traceChild = <T>(name: string, callback: TraceCallback<T>, metadata?: TraceMetadata) => {
+    const child = transaction?.startChild({ ...metadata, op: name })
+    return traceTransaction(child)(name, callback)
+  }
+  const setTraceData = (key: string, value: unknown) => void transaction?.setData(key, value)
```

Comment:
subjective nit: braces make these easier to read

Reply:
I have a strong preference for one-liners (probably from C++), so I often can't tell. Adding brackets :)